FrostCord/FrostCord
# Role Assignment and Other Optimizations

# Role Assignment + Optimiz
This PR deprecates the `UserPermsState` and `useUserPermsStore` interface/store respectively

# API Changes
- Introduces a double `Map` based server profile cache for storing/maintaining server user profiles for use throughout servers. The structure is basically `server_id -> profile_id -> profile` where the first map, indexed by `server_id` has all the server profiles for users in that server, accessible via their profile id.
- This has also brought the following changes:
  - More reliable user permissions checking and role heirarchy checking
  - Cleaner messaging struct as we no longer need to fetch for a `MessageWithServerProfile`, but instead can rely on the cache
  - Real-time user profile updates for all manner of events such as role assignment/removal, role creation/deletion, nickname changes, avatar changes, etc.
- I have also added the messages tab back for the mobile view
- Miniprofiles now open on click, addressing a prior UX issue that was discussed

# Testing
- Verify roles can be created/assigned if you have the permission to manage roles
- Verify you can't remove roles from users that are above your current highest role
- Verify that adding a role with a color updates message headers for that user in realtime wth the highest role color
- Verify that removing it reverts the message header color to that of theri current highest role
- Verify that assigning a user a role and then deleting the role from the server will be reflected in their mini profile/message header
- Verify DMs still work and fetch profile info appropriately on new DM and when navigating existing ones
- Verify roles reflect their permissions in real time:
  - Assign a role to a user, give it manage messages
  - Verify the user can now manage messages in the server
  - Remove the permission to manage messages from the role and save it, verify the user can no longer manage messages
  - Add the permission back, verify they can now manage messages again
  - Delete the role from the server, verify they can no longer manage messages
